hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/HTSPrecompiledContract.java

```diff
@@ -493,6 +529,75 @@ TypedTokenStore newTokenStore(
 		}
 	}
 
+	private class TransferPrecompile implements Precompile {
+		private SyntheticTxnFactory.FungibleTokenTransfer transferOp;
+
+		@Override
+		public TransactionBody.Builder body(final Bytes input) {
+			transferOp = decoder.decodeTransferToken(input);
+			return syntheticTxnFactory.createCryptoTransfer(List.of(), List.of(), List.of(transferOp));
+		}
+
+		@Override
+		public ExpirableTxnRecord.Builder run(
+				final Address recipient,
+				final Address contract,
+				final WorldLedgers ledgers
+		) {
+			final List<BalanceChange> changes = List.of(
+					BalanceChange.changingFtUnits(
+							Id.fromGrpcToken(transferOp.getDenomination()),
+							transferOp.getDenomination(),
+							AccountAmount.newBuilder().setAccountID(transferOp.sender).setAmount(-transferOp.amount).build()
+					),
+					BalanceChange.changingFtUnits(
+							Id.fromGrpcToken(transferOp.getDenomination()),
+							transferOp.getDenomination(),
+							AccountAmount.newBuilder().setAccountID(transferOp.receiver).setAmount(transferOp.amount).build()
+					)
+			);
+
+			var validated = impliedTransfersMarshal.assessCustomFeesAndValidate(
+					changes,
+					changes.size(),
+					impliedTransfersMarshal.currentProps()
+			);
+
+			final var sideEffects = sideEffectsFactory.get();
+			final var hederaTokenStore = hederaTokenStoreFactory.newHederaTokenStore(
+					ids,
+					validator,
+					sideEffects,
+					NOOP_VIEWS_MANAGER,
+					dynamicProperties,
+					ledgers.tokenRels(), ledgers.nfts(), ledgers.tokens());
+			hederaTokenStore.setAccountsLedger(ledgers.accounts());
+
+
+			final var transferLogic = transferLogicFactory.newLogic(
+					ledgers.accounts(), ledgers.nfts(), ledgers.tokenRels(), hederaTokenStore,
+					sideEffects,
+					NOOP_VIEWS_MANAGER,
+					dynamicProperties,
+					validator);
+
+			var solidityAddressFrom = EntityIdUtils.asTypedSolidityAddress(transferOp.sender);
+			var solidityAddressTo = EntityIdUtils.asTypedSolidityAddress(transferOp.receiver);
+
+			final var hasRequiredSigs = sigsVerifier.hasActiveKeyOrNoReceiverSigReq(
```

Comment:
As noted above, the `List<BalanceChange>` list may include many more than 2 items once custom fees have been assessed.

So we need a loop that checks the signing requirements for **all** the changes, like,
```
for (final var change : changes) {
    final var units = change.units();
    if (units < 0) {
        final var hasSenderSig = sigsVerifier.hasActiveKey(change.account(), recipient, contract);
        validateTrue(hasSenderSig, INVALID_SIGNATURE);
    } else if (units > 0) {
        /* Need to add the Id.asSolidityAddress() method. */
        final var hasReceiverSigIfReq = sigsVerifier.hasActiveKeyOrNoReceiverSigReq(id.asSolidityAddress(), recipient, contract);
        validateTrue(hasReceiverSigIfReq, INVALID_SIGNATURE);
    }
}
``` 
Note that the 2nd and 3rd arguments to any `sigsVerifier` method are always the same.